taxcalc/filings/forms/tax_form.py

```diff
@@ -0,0 +1,179 @@
+        If not specified defaults to zero.
+        """
+        return self._fields.get(self.__class__._TAX_UNIT_ID_FIELD) or '0'
+
+    def set_field(self, name, value):
+        """
+        Helper method for setting a single field.
+        """
+        self.set_fields({name: value})
+
+    def set_fields(self, data):
+        """
+        Set given fields if they pass validation.
+        """
+        if not isinstance(data, dict):
+            raise ValueError('Expected a dict of field data')
+        valid_fields = self.__class__._VALID_FIELDS
+        for key, value in data.items():
+            if valid_fields and key not in valid_fields:
+                print('Unknown field {0} for form {2}'.
+                      format(key, self.form_name()))
+            else:
+                self._fields[key] = value
+
+    def to_evars(self):
+        """
+        Get stored data as a dict of evars.
+        Raises an error on conflicting calculations.
+        Converts field values (stored as strings) to floats
+        """
+        direct = self.to_evars_direct()
+        indirect = self.to_evars_indirect() or {}
+
+        # Check for conflicts
+        for key in direct:
+            if key in indirect and direct[key] != indirect[key]:
+                raise ValueError('Different calc for same evar.')
+        for key in indirect:
+            if key in direct and direct[key] != indirect[key]:
+                raise ValueError('Different calc for same evar.')
+
+        direct.update(indirect)
+        return direct
+
+    def to_evars_direct(self):
+        """
+        Get stored data as a dict of evars following defined evar maps.
+        Raises an error on conflicting calculations.
+        Converts field values stored as strings to int/float.
+        """
+        results = {}
+
+        if self._EVAR_MAP_BY_YEAR:
+            year_evar_map = self._EVAR_MAP_BY_YEAR[self.year]
+        else:
+            year_evar_map = None
+
+        for key, value in self._fields.items():
+            if self._EVAR_MAP and key in self._EVAR_MAP:
+                evar = self._EVAR_MAP[key]
+            elif year_evar_map and key in year_evar_map:
+                evar = year_evar_map[key]
+            else:
+                continue
+            if evar in results and results[evar] != value:
+                raise ValueError('Different calc for same evar.')
+            results[evar] = value
+
+        # convert results to value types
+        for key, value in results.items():
+            if isinstance(value, six.string_types):
+                results[key] = string_to_number(value)
+
+        return results
+
+    def to_evars_indirect(self):
+        """
+        Get any evars that can't be determined with a simple 1-1 mapping,
+        e.g. if an evar depends on multiple lines.
+        To be extended by child classes.
+        Should convert values to their expected types.
+        """
+
```

Comment:
We don't need a `raise NotImplementedError`, because we expect this function to produce `None` if a child TaxForm class has no indirect evar mappings.

I believe we don't need a `pass` because the function has a comment block, thus avoiding the `IndentationError` that a function without any body content would produce. If we added a `pass` here, might we encounter pylint message W0107 - "Unneccesary pass statement"? If not and you believe the pass statement adds clarity I'm happy to add it.